### PR TITLE
implement ArcSort.

### PR DESCRIPTION
**Pull request by csukuangfj**

### CODE REVIEW

**Comment by csukuangfj** [comment]
The non-in-place version will be added after #217  is resolved.

**Comment by danpovey** [comment]
Looks good to me!  One of u guys can merge if you think it's ready.

**Comment by csukuangfj** [comment]
I'll merge it after fixing `FsaFromString`: Sort the src_states only if `openfst == true`.

**Comment by csukuangfj** [comment]
merging.

File: k2/csrc/ragged_inl.h [hunk 1]
```diff
@@ -103,23 +103,35 @@ Ragged<T> RandomRagged(T min_value, T max_value, int32_t min_num_axes,
...
+                                 *context);           // context
+  else
+    mgpu::segmented_sort(src->values.Data(),  // keys
+                         src->values.Dim(),   // count
+                         segment.Data(),      // segments
+                         segment.Dim() - 1,   // num_segments
+                         Op(),                // cmp
+                         *context);           // context
+
   auto err = cudaGetLastError();
```
Comment (qindazhu) [inline k2/csrc/ragged_inl.h:135]: wondering if there's still cuda error now?

Comment (csukuangfj) [reply k2/csrc/ragged_inl.h:135]: It disappears!

Comment (qindazhu) [reply k2/csrc/ragged_inl.h:135]: ok, then i guess the error before was from some other code (and fixed now) because we don't sync kernel call

**Pull request merged**